### PR TITLE
to_dense and to_sparse

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #58

Differential Revision: [D36682857](https://our.internmc.facebook.com/intern/diff/D36682857)

### CODE REVIEW

File: maskedtensor/core.py [hunk 1]
```diff
@@ -87,6 +87,50 @@ def backward(ctx, grad_output):
         return grad_output
 
 
+class MaskedToDense(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input):
+        ctx.save_for_backward(input)
```
Comment (cpuhrsch) [inline maskedtensor/core.py:93]: It seems much cheaper to only save the `input.layout`

File: maskedtensor/core.py [hunk 2]
```diff
@@ -87,6 +87,50 @@ def backward(ctx, grad_output):
...
+            return grad_output.to_sparse()
+        elif input.layout == torch.strided:
+            return grad_output.to_dense()
+        raise ValueError("to_dense: Unsupported input layout: ", input.layout)
+
+
+class MaskedToSparse(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input):
+        ctx.save_for_backward(input)
```
Comment (cpuhrsch) [inline maskedtensor/core.py:116]: You don't need to save this at all based on the code in backward.

File: maskedtensor/core.py [hunk 3]
```diff
@@ -338,6 +386,55 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
             new_data = func(args[0], get_data(mx), get_data(my))
             new_mask = func(args[0], get_mask(mx), get_mask(my))
+            return MaskedTensor(new_data, new_mask)
+        if func is torch.ops.aten.to_sparse:
+            assert len(args) == 1
+            assert len(kwargs) == 0
+            assert torch.is_tensor(args[0])
+            mt = args[0]
+            if not is_masked_tensor(mt):
+                mt = MaskedTensor(mt, torch.ones_like(mt).bool())
```
Comment (cpuhrsch) [inline maskedtensor/core.py:396]: When does this happen? If there's only 1 arg torch_dispatch shouldn't be called for MaskedTensor unless one of the tensors is a MaskedTensor. If this does happen maybe it's a bug report for torch_dispatch?

File: maskedtensor/core.py (hunk 1 above)
Comment (george-qi) [reply maskedtensor/core.py:93]: Apparently we're not allowed to save only the layout

File: maskedtensor/core.py (hunk 3 above)
Comment (george-qi) [reply maskedtensor/core.py:396]: This gets called in the backwards function

```
converted = mt.to_sparse().to_dense()
converted.sum().backward()
```

Printing out when we enter `__torch_function__` and `__torch_dispatch__`:

```
torch_function <method 'to_sparse' of 'torch._C._TensorBase' objects>
torch_function <method-wrapper '__get__' of getset_descriptor object at 0x7f2430839c00>
torch_function <method 'to_dense' of 'torch._C._TensorBase' objects>
torch_function <method-wrapper '__get__' of getset_descriptor object at 0x7f2430839c00>
** end of first line
torch_function <method 'sum' of 'torch._C._TensorBase' objects>
torch_function <function Tensor.backward at 0x7f24306caca0>
torch_dispatch aten.ones_like
torch_dispatch aten.to_sparse -- here
torch_dispatch aten._to_dense -- and here
torch_dispatch aten.detach
torch_dispatch aten.detach
** end of second line
```

File: maskedtensor/core.py (hunk 1 above)
Comment (cpuhrsch) [reply maskedtensor/core.py:93]: Oh, I see. We probably need to extend save_for_backward to support Layout. A workaround here is mapping layouts to integers and saving those.

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36682857).

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36682857).

**Pull request closed without merging**